chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,248 @@
+
+            _dist_term_sample = F.sum(
+                F.maximum(cp.asarray(0.0, np.float32), margin - nrm) ** 2)
+            dist_term += _dist_term_sample / cp.asarray((nobj * (nobj - 1)),
+                                                        np.float32)
+
+        dist_term = dist_term / cp.asarray(bs, dtype=np.float32)
+
+        return dist_term
+
+    def regularization_term(self, means, n_objects):
+        """
+        means: bs, n_instances, n_filters
+        """
+
+        bs, n_instances, n_filters = means.shape
+        m = F.cast(means, np.float32)
+
+        reg_term = 0.0
+        for i in range(bs):
+            reg_term += F.mean(self.norm(m[i, : n_objects[i], :], 1))
+        reg_term = reg_term / cp.asarray(bs, dtype=np.float32)
+
+        return reg_term
+
+    def means(self, pred, gt, n_objects, max_n_objects, gt_idx):
+        """pred: bs, n_filters, height * width
+           gt: bs, n_instances, height * width"""
+
+        bs = pred.shape[0]
+        n_filters = pred.shape[1]
+        n_instances = gt.shape[1]
+        n_loc = pred.shape[2]
+
+        # Expand prediction and broadcast instances to 3rd axis
+        p = F.reshape(pred, (bs, n_filters, 1, n_loc))
+        p = F.broadcast_to(p, (bs, n_filters, n_instances, n_loc))
+
+        # Expand ground truth to match the size but do not broadcast
+        g = F.reshape(gt, (bs, 1, n_instances, n_loc))
+
+        p = p * F.cast(g, p.dtype)
+
+        means = []
+        for i in range(bs):
+
+            p_item = p[i, :, gt_idx[i]]
+            g_item = g[i, :, gt_idx[i]]
+
+            p_sum = F.sum(p_item, axis=2)
+            g_sum = F.cast(F.sum(g_item, axis=2), p_sum.dtype)
+            _mean_sample = p_sum / g_sum
+
+            n_fill_objects = max_n_objects - n_objects[i]
+
+            assert n_fill_objects >= 0
+
+            if n_fill_objects != 0:
+                _fill_sample = cp.zeros(
+                    (cp.asnumpy(n_fill_objects), n_filters),
+                    dtype=_mean_sample.dtype)
+                _mean_sample = F.concat((_mean_sample, _fill_sample), axis=0)
+
+            means.append(_mean_sample)
+
+        means = F.stack(means)
+        return means
+
+    def prepare_inputs(self, prediction, labels):
+        # Reshape layers to prepare for processing
+        p_shape = prediction.shape
+        l_shape = labels.shape
+
+        prediction = F.reshape(prediction, shape=(
+            p_shape[0], p_shape[1], p_shape[2] * p_shape[3]))
+        labels = F.reshape(labels, shape=(
+            l_shape[0], l_shape[1], l_shape[2] * l_shape[3]))
+
+        return prediction, labels
+
+    def apply(self, *x):
+        """
+        :param x:
+        x[0] = segmentation prediction output
+        x[1] = segmentation ground truth
+        x[2] = number of objects in ground truth
+        x[3] = indexes of non-zero ground truths
+        :return:
+        Loss value
+        """
+
+        assert len(x) == 4
```

Comment:
If we do not use variable-length argument, we do not have to check the number of arguments with assertion.